django/db/backends/postgresql_psycopg2/operations.py

```diff
@@ -45,13 +47,16 @@ def datetime_trunc_sql(self, lookup_type, field_name, tzname):
     def deferrable_sql(self):
         return " DEFERRABLE INITIALLY DEFERRED"
 
-    def lookup_cast(self, lookup_type):
+    def lookup_cast(self, lookup_type, internal_type=None):
         lookup = '%s'
 
         # Cast text lookups to text to allow things like filter(x__contains=4)
         if lookup_type in ('iexact', 'contains', 'icontains', 'startswith',
                            'istartswith', 'endswith', 'iendswith', 'regex', 'iregex'):
-            lookup = "%s::text"
+            if internal_type in ('IPAddressField', 'GenericIPAddressField'):
+                lookup = "HOST(%s)"
```

Comment:
I'll leave it be for now, I intend to deprecate support for text based operations on non text fields in Django 1.9 anyway, and replace with a `__text` transform (which would be overridden for postgres on the `GenericIPAddressField` to do `HOST()` instead of `::text`).